cpp/include/cudf/fixed_point/fixed_point.hpp

```diff
@@ -173,6 +224,153 @@ CUDF_HOST_DEVICE inline constexpr T shift(T const& val, scale_type const& scale)
+ *
+ * @note This returns all zero's for +/-0 and denormals, and zero's with an exponent of
+ * INT_MIN for the unrepresentable values +/-inf and NaN's.
+ *
+ * @tparam FloatingType Type of input floating point value
+ * @param floating The floating value to extract the components from
+ * @return A tuple containing the sign (bool), exponent (power-of-2), and integral significand.
+ */
+template <typename FloatingType,
+          typename cuda::std::enable_if_t<cuda::std::is_floating_point_v<FloatingType>>* = nullptr>
+CUDF_HOST_DEVICE inline auto extract_components(FloatingType floating)
+{
+  // This function is implemented using integer and bit operations, which is much
+  // faster than using the FPU functions for each component individually. This is especially
+  // true for doubles, as bottlenecking on the FP64 GPU pipeline can cut performance in half.
+
+  // See comments in FloatingComponentConstants about the floating point format.
+  using Constants    = FloatingComponentConstants<FloatingType>;
+  using IntegralType = typename Constants::IntegralType;
+
+  // Convert floating to integer
+  auto const integer_rep = [&]() {
+    IntegralType integer;
+    memcpy(&integer, &floating, sizeof(floating));
+    return integer;
+  }();
+
+  // First extract the exponent bits and handle its special values
+  auto const exponent_bits = integer_rep & Constants::exponent_mask;
+  if (exponent_bits == 0) {
+    // Because of the understood set-bit not stored in the mantissa, it is not possible
+    // to store the value zero directly. Instead both +/-0 and denormals are represented with
+    // the exponent bits set to zero.
+    // Thus it's fastest to just floor (generally unwanted) denormals to zero.
+    return cuda::std::tuple{false, 0, IntegralType(0)};
+  } else if (exponent_bits == Constants::exponent_mask) {
+    //+/-inf and NaN values are stored with all of the exponent bits set.
+    // As none of these are representable by integers, we'll return the same value for all cases.
+    return cuda::std::tuple{false, INT_MIN, IntegralType(0)};
+  }
+
+  // Extract the exponent value: shift the bits down and subtract the bias.
+  using SignedIntegralType                       = cuda::std::make_signed_t<IntegralType>;
+  SignedIntegralType const shifted_exponent_bits = exponent_bits >> Constants::num_mantissa_bits;
+  int const exp2 =
+    shifted_exponent_bits - static_cast<SignedIntegralType>(Constants::exponent_bias);
+
+  // Extract the sign bit:
+  bool const is_positive = ((Constants::sign_mask & integer_rep) == 0);
+
+  // Extract the significand, setting the high bit for the understood 1/2
+  IntegralType const significand =
+    (integer_rep & Constants::mantissa_mask) | Constants::understood_bit_mask;
+
+  // Return the tuple of values
+  return cuda::std::tuple{is_positive, exp2, significand};
+}
+
+/** @brief Sets the sign and adds to the exponent of a floating point number.
```

Comment:
Question: why would you need to set the sign bit of a float explicitly? Why not
```
   float out = (is_negative) ? -fabs(in) : fabs(in)
```

Question: what are you adding to the exponent, and why? And why is this combined with setting the sign bit rather than a separate function?

Reply:
Basically, when we convert decimal --> floating, what we're really doing is converting a decimal in base 10 (i.e. with a base 10 scale factor) to a decimal in base 2 (i.e. with a base 2 scale factor).  Then we cast this base-2 decimal value to a float/double, and then store the base-2 scale factor within the float/double.  So this add of exp2 is the storing of the base-2 scale factor in the floating point number.  This conversion will be in the main, upcoming PR.  

As far as the sign bit goes, setting the bit this way avoids a branch.

Reply:
OK, so why not have separate `set_sign_bit` and `add_exp2` functions? Currently the name is confusing (it doesn't add the sign bit, it sets it).  And as you explain above, you are storing/setting the base-2 scale factor,  not adding it. So maybe `set_base2_scale_factor` would be a better name.

Reply:
That would be 2x the memcpy's, it's faster to do them both at once.  This is really a single step of "finalizing" the conversion to floating point, these two operations won't ever be done separately.  

I'm actually adding to the base-2 scale factor instead of setting it, as after the cast of some integer (e.g. 12345678) to float will already have a base-2 exponent built-in to the floating point value.  I'll change the name of the function to set_sign_add_exp2()